File: src/main/java/com/exception_study/api/product/service/ProductService.java
package com.exception_study.api.product.service;

import com.exception_study.api.product.dto.ProductDto;
import com.exception_study.api.product.dto.response.DetailsWithHistoryResponse;
import com.exception_study.api.product.entity.Product;
import com.exception_study.api.product.repository.ProductRepository;
import com.exception_study.api.product_order.dto.ProductOrderDto;
import com.exception_study.api.user_account.dto.request.RegisterRequest;
import com.exception_study.api.user_account.entity.UserAccount;
import com.exception_study.global.exception.ErrorCode;
import com.exception_study.global.exception.StudyApplicationException;
import com.exception_study.api.product_order.repository.ProductOrderRepository;
import com.exception_study.api.user_account.repository.UserAccountRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Transactional
@RequiredArgsConstructor
@Service
public class ProductService {

    private final ProductRepository productRepository;
    private final UserAccountRepository userAccountRepository;
    private final ProductOrderRepository productOrderRepository;

    @Transactional(readOnly = true)
    public List<ProductDto> getList() {
        return productRepository.findAll().stream().map(ProductDto::from).toList();
    }

    public void register(RegisterRequest dto, String userId) {
        UserAccount userAccount = userAccountRepository.findById(userId).orElseThrow(
                () -> new StudyApplicationException(ErrorCode.USER_NOT_FOUND)
        );
        Product product = Product.of(dto.getName(), dto.getPrice(),dto.getQuantity(),userAccount);
        productRepository.save(product);
    }

    @Transactional(readOnly = true)
    public ProductDto getDetails(Long id) {
        Product product = productRepository.findById(id).orElseThrow(
                () -> new StudyApplicationException(ErrorCode.PRODUCT_NOT_FOUND)
        );
        return ProductDto.from(product);
    }

    @Transactional(readOnly = true)
    public DetailsWithHistoryResponse getDetails(Long id, String userId) {
        UserAccount userAccount = userAccountRepository.findById(userId).orElseThrow(
                () -> new StudyApplicationException(ErrorCode.USER_NOT_FOUND)
        );
        Product product = productRepository.findById(id).orElseThrow(
                () -> new StudyApplicationException(ErrorCode.PRODUCT_NOT_FOUND)
        );

        List<ProductOrderDto> history = productOrderRepository.findTransaction_History(product.getSeller().getUserId(),userAccount.getUserId())
                .stream().map(ProductOrderDto::from).toList();

        return DetailsWithHistoryResponse.of(ProductDto.from(product),history);
    }


}

File: src/product/product.service.ts
import {
  Injectable,
  InternalServerErrorException,
  NotFoundException,
  UnauthorizedException,
} from '@nestjs/common';
import { ProductRepository } from './product.repository';
import { Product } from './entity/product.entity';
import { ProductDto } from './dto/product.dto';

@Injectable()
export class ProductService {
  constructor(private readonly productRepository: ProductRepository) {}

  // 제품 목록 조회
  async findAllProducts(): Promise<Product[]> {
    return await this.productRepository.findAllProducts();
  }

  // 제품 상세 조회
  async findProductDetail(productId: number): Promise<Product> {
    const product = await this.productRepository.findProductDetail(productId);
    if (!product) {
      throw new NotFoundException('존재하지 않는 상품입니다.');
    }
    return product;
  }

  // 제품 올리기
  async createProduct(
    userId: number,
    productDto: ProductDto,
  ): Promise<Product> {
    const product = await this.productRepository.createProduct(
      userId,
      productDto,
    );
    if (!product) {
      throw new InternalServerErrorException('제품 올리기 실패');
    }
    return product;
  }

  // 제품 판매 승인
  async approveProductSale(userId: number, productId: number): Promise<void> {
    const product = await this.findProductDetail(productId);
    if (product.userId !== userId) {
      throw new UnauthorizedException('제품 판매 승인 권한이 없습니다.');
    }
    await this.productRepository.approveProductSale(productId);
  }

  // 거래내역
  async findTransactionHistory(
    userId: number,
    sellerId: number,
  ): Promise<Product[]> {
    return await this.productRepository.findTransactionHistory(
      userId,
      sellerId,
    );
  }
}
